src/rgw/rgw_flight.h

```diff
@@ -0,0 +1,222 @@
+    FlightData(const FlightData&) = default;
+    FlightData(FlightData&&) = default;
+    FlightData& operator=(const FlightData&) = default;
+    FlightData& operator=(FlightData&&) = default;
+  };
+
+  // stores flights that have been created and helps expire them
+  class FlightStore {
+
+  protected:
+
+    boost::intrusive_ptr<ceph::common::CephContext> cct;
+    const DoutPrefix dp;
+
+  public:
+
+    FlightStore(boost::intrusive_ptr<ceph::common::CephContext>& _cct);
+    virtual ~FlightStore();
+    virtual FlightKey add_flight(FlightData&& flight) = 0;
+
+    // TODO consider returning const shared pointers to FlightData in
+    // the following two functions
+    virtual arw::Result<FlightData> get_flight(const FlightKey& key) const = 0;
+    virtual std::optional<FlightData> after_key(const FlightKey& key) const = 0;
+
+    virtual int remove_flight(const FlightKey& key) = 0;
+    virtual int expire_flights() = 0;
+  };
+
+  class MemoryFlightStore : public FlightStore {
+    std::map<FlightKey, FlightData> map;
+    mutable std::mutex mtx; // for map
+
+  public:
+
+    MemoryFlightStore(boost::intrusive_ptr<ceph::common::CephContext>& _cct);
+    virtual ~MemoryFlightStore();
+    FlightKey add_flight(FlightData&& flight) override;
+    arw::Result<FlightData> get_flight(const FlightKey& key) const override;
+    std::optional<FlightData> after_key(const FlightKey& key) const override;
+    int remove_flight(const FlightKey& key) override;
+    int expire_flights() override;
+  };
+
+  class FlightServer : public flt::FlightServerBase {
+
+    using Data1 = std::vector<std::shared_ptr<arw::RecordBatch>>;
+
+    boost::intrusive_ptr<ceph::common::CephContext> cct;
+    const DoutPrefix dp;
+    rgw::sal::Store* store;
+    FlightStore* flight_store;
+
+    std::map<std::string, Data1> data;
+
+  public:
+
+    static constexpr int default_port = 8077;
+
+    FlightServer(boost::intrusive_ptr<ceph::common::CephContext>& _cct,
+		 rgw::sal::Store* _store,
+		 FlightStore* _flight_store);
+    ~FlightServer() override;
+
+    FlightStore* get_flight_store() {
+      return flight_store;
+    }
+
+    arw::Status ListFlights(const flt::ServerCallContext& context,
+			    const flt::Criteria* criteria,
+			    std::unique_ptr<flt::FlightListing>* listings) override;
+
+    arw::Status GetFlightInfo(const flt::ServerCallContext &context,
+			      const flt::FlightDescriptor &request,
+			      std::unique_ptr<flt::FlightInfo> *info) override;
+
+    arw::Status GetSchema(const flt::ServerCallContext &context,
+			  const flt::FlightDescriptor &request,
+			  std::unique_ptr<flt::SchemaResult> *schema) override;
+
+    arw::Status DoGet(const flt::ServerCallContext &context,
+		      const flt::Ticket &request,
+		      std::unique_ptr<flt::FlightDataStream> *stream) override;
+  }; // class FlightServer
+
+  // class OwningStringView : public std::string_view {
+  class OwningStringView : public arw::util::string_view {
```

Comment:
isn't an owning string view just a `std::string`? could that could work here?

Reply:
I just looked into this and there are a couple of issues. One is that there's a Peek function in the user interface that needs to return a *their* subclass of std::string_view. Also, we want to pre-allocate the buffer that the data is being copied into with minimal fuss, as OwngingStringView does. To create a std:string with a pre-allocated buffer of a given size, we need to copy data into it, and the effort to copy would be wasted.

But I appreciate the thoughtful comments as it was helpful to revisit the decisions here.